#### We will use pima indian diabetes dataset to predict if a person has a diabetes or not based on certain features such as blood pressure, skin thickness, age etc. We will train a standalone model first and then use bagging ensemble technique to check how it can improve the performance of the model

dataset credit: https://www.kaggle.com/gargmanas/pima-indians-diabetes


In [1]:
import pandas as pd

In [3]:
df = pd.read_csv('/Users/hirdeshpal/Downloads/diabetes.csv')
df.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   
5            5      116             74              0        0  25.6   
6            3       78             50             32       88  31.0   
7           10      115              0              0        0  35.3   
8            2      197             70             45      543  30.5   
9            8      125             96              0        0   0.0   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
5                     0.201   30        0  
6                     0.248   26        1  
7                     0.134   29        0  
8                     0.158   53        1  
9                     0.232   54        1

In [5]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [7]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [19]:
df.Outcome.value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

There is slight imbalance in our dataset but since it is not major we will not worry about it!

In [25]:
X = df.drop('Outcome', axis = 1)
y = df.Outcome

## Scaling the data

In [31]:
from sklearn.preprocessing import StandardScaler

In [37]:
# Scaling the dataset to correct the units of all the columns
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
print(X_scaled)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [44]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, stratify=y, random_state=10)
# stratify=y: is used to maintain train and test imbalance to some extant

## Train using stand alone model

In [48]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

In [54]:
score = cross_val_score(DecisionTreeClassifier(),X,y,cv=5)
score

array([0.65584416, 0.68181818, 0.72727273, 0.78431373, 0.73856209])

In [56]:
score.mean()

0.7175621763857057

## Train using Bagging

In [64]:
from sklearn.ensemble import BaggingClassifier

In [68]:
bag_model = BaggingClassifier(
                          estimator=DecisionTreeClassifier(), #which model want to use
                          n_estimators=100, # Number model need to create
                          max_samples=0.8,  # 80% data used
                          oob_score=True,
                          random_state=0
                         )
bag_model.fit(X_train,y_train)
bag_model.oob_score_

0.7534722222222222

In [70]:
bag_model.score(X_test,y_test)

0.7760416666666666

In [72]:
bag_model = BaggingClassifier(
                          estimator=DecisionTreeClassifier(), #which model want to use
                          n_estimators=100, # Number model need to create
                          max_samples=0.8,  # 80% data used
                          oob_score=True,
                          random_state=0
                         )
score = cross_val_score(bag_model, X, y, cv=5)
score.mean()

0.7578728461081402

## Train using Random Forest

In [78]:
from sklearn.ensemble import RandomForestClassifier 

In [80]:
score = cross_val_score(RandomForestClassifier(),X,y,cv = 5)
score.mean()

0.7656990068754774

# Exercise

Download heart disease dataset heart.csv in Exercise folder and do following, (credits of dataset: https://www.kaggle.com/fedesoriano/heart-failure-prediction)

1. Load heart disease dataset in pandas dataframe
2. Remove outliers using Z score. Usual guideline is to remove anything that has Z score > 3 formula or Z score < -3
3. Convert text columns to numbers using label encoding and one hot encoding
4. Apply scaling
5. Build a classification model using support vector machine. Use standalone model as well as Bagging model and check if you see any difference in the performance.
6. Now use decision tree classifier. Use standalone model as well as Bagging and check if you notice any difference in performance
7. Comparing performance of svm and decision tree classifier figure out where it makes most sense to use bagging and why. Use internet to figure out in what conditions bagging works the best.

### 1. Load heart disease dataset in pandas dataframe

In [88]:
import pandas as pd

In [90]:
df = pd.read_csv('/Users/hirdeshpal/Downloads/heart.csv')
df.head(10)

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   
5   39   M           NAP        120          339          0     Normal    170   
6   45   F           ATA        130          237          0     Normal    170   
7   54   M           ATA        110          208          0     Normal    142   
8   37   M           ASY        140          207          0     Normal    130   
9   48   F           ATA        120          284          0     Normal    120   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0  
5              N      0.0       Up             0  
6              N      0.0       Up             0  
7              N      0.0       Up             0  
8              Y      1.5     Flat             1  
9              N      0.0       Up             0

In [96]:
df.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   198.799564    0.233115  136.809368   
std      9.432617   18.514154   109.384145    0.423046   25.460334   
min     28.000000    0.000000     0.000000    0.000000   60.000000   
25%     47.000000  120.000000   173.250000    0.000000  120.000000   
50%     54.000000  130.000000   223.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

In [98]:
df.shape

(918, 12)

In [100]:
df.isnull().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

### 2. Remove outliers using Z score. Anything that has Z score > 3 formula or Z score < -3

In [107]:
df[df.Cholesterol>  (df.Cholesterol.mean() + 3*df.Cholesterol.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
76    32   M           ASY        118          529          0     Normal   
149   54   M           ASY        130          603          1     Normal   
616   67   F           NAP        115          564          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
76     130              N      0.0     Flat             1  
149    125              Y      1.0     Flat             1  
616    160              N      1.6     Flat             0

In [109]:
df.shape

(918, 12)

In [111]:
df1 = df[df.Cholesterol<=  (df.Cholesterol.mean() + 3*df.Cholesterol.std())]
df1.shape

(915, 12)

In [113]:
df[df.MaxHR > (df.MaxHR.mean()+3*df.MaxHR.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [115]:
df[df.FastingBS>(df.FastingBS.mean()+3*df.FastingBS.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [117]:
df[df.Oldpeak>(df.Oldpeak.mean()+3*df.Oldpeak.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
166   50   M           ASY        140          231          0         ST   
702   59   M            TA        178          270          0        LVH   
771   55   M           ASY        140          217          0     Normal   
791   51   M           ASY        140          298          0     Normal   
850   62   F           ASY        160          164          0        LVH   
900   58   M           ASY        114          318          0         ST   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
166    140              Y      5.0     Flat             1  
702    145              N      4.2     Down             0  
771    111              Y      5.6     Down             1  
791    122              Y      4.2     Flat             1  
850    145              N      6.2     Down             1  
900    140              N      4.4     Down             1

In [119]:
df2 = df1[df1.Oldpeak<=(df1.Oldpeak.mean()+3*df1.Oldpeak.std())]
df2.shape

(909, 12)

In [121]:
df[df.RestingBP>(df.RestingBP.mean()+3*df.RestingBP.std())]


Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
109   39   M           ATA        190          241          0     Normal   
241   54   M           ASY        200          198          0     Normal   
365   64   F           ASY        200            0          0     Normal   
399   61   M           NAP        200            0          1         ST   
592   61   M           ASY        190          287          1        LVH   
732   56   F           ASY        200          288          1        LVH   
759   54   M           ATA        192          283          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
109    106              N      0.0       Up             0  
241    142              Y      2.0     Flat             1  
365    140              Y      1.0     Flat             1  
399     70              N      0.0     Flat             1  
592    150              Y      2.0     Down             1  
732    133              Y      4.0     Down             1  
759    195              N      0.0       Up             1

In [123]:
df3 = df2[df2.RestingBP<=(df2.RestingBP.mean()+3*df2.RestingBP.std())]
df3.shape

(902, 12)

In [125]:
df.ChestPainType.unique()

array(['ATA', 'NAP', 'ASY', 'TA'], dtype=object)

In [127]:
df.RestingECG.unique()


array(['Normal', 'ST', 'LVH'], dtype=object)

In [129]:
df.ExerciseAngina.unique()


array(['N', 'Y'], dtype=object)

In [131]:
df.ST_Slope.unique()


array(['Up', 'Flat', 'Down'], dtype=object)

### 3. Handle text columns using label encoding and one hot encoding

In [134]:
df4 = df3.copy()

In [140]:
df4.ExerciseAngina.replace({
    'N':0,
    'Y':1
},inplace=True)
df4.ST_Slope.replace(
    {
        'Down': 1,
        'Flat': 2,
        'Up': 3
    },
    inplace=True
)

df4.RestingECG.replace(
    {
        'Normal': 1,
        'ST': 2,
        'LVH': 3
    },
    inplace=True)

df4.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40   M           ATA        140          289          0           1   
1   49   F           NAP        160          180          0           1   
2   37   M           ATA        130          283          0           2   
3   48   F           ASY        138          214          0           1   
4   54   M           NAP        150          195          0           1   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172               0      0.0         3             0  
1    156               0      1.0         2             1  
2     98               0      0.0         3             0  
3    108               1      1.5         2             1  
4    122               0      0.0         3             0

In [144]:
df5 = pd.get_dummies(df4,drop_first=True).astype(int)
df5.head()

Age  RestingBP  Cholesterol  FastingBS  RestingECG  MaxHR  ExerciseAngina  \
0   40        140          289          0           1    172               0   
1   49        160          180          0           1    156               0   
2   37        130          283          0           2     98               0   
3   48        138          214          0           1    108               1   
4   54        150          195          0           1    122               0   

   Oldpeak  ST_Slope  HeartDisease  Sex_M  ChestPainType_ATA  \
0        0         3             0      1                  1   
1        1         2             1      0                  0   
2        0         3             0      1                  1   
3        1         2             1      0                  0   
4        0         3             0      1                  0   

   ChestPainType_NAP  ChestPainType_TA  
0                  0                 0  
1                  1                 0  
2                  0                 0  
3                  0                 0  
4                  1                 0

## 4. Apply scaling

In [154]:
X = df5.drop('HeartDisease',axis=1)
y = df5.HeartDisease

In [156]:
from sklearn.preprocessing import StandardScaler

In [158]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled

array([[-1.42896269,  0.46089071,  0.85238015, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-0.47545956,  1.5925728 , -0.16132855, ..., -0.4836591 ,
         1.86750159, -0.22914788],
       [-1.74679706, -0.10495034,  0.79657967, ...,  2.06757196,
        -0.53547478, -0.22914788],
       ...,
       [ 0.37209878, -0.10495034, -0.61703246, ..., -0.4836591 ,
        -0.53547478, -0.22914788],
       [ 0.37209878, -0.10495034,  0.35947592, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-1.64085227,  0.3477225 , -0.20782894, ..., -0.4836591 ,
         1.86750159, -0.22914788]])

## 5. Train a model using standalone support vector machine and then using bagging

In [163]:
from sklearn.model_selection import train_test_split

In [165]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled,y,test_size=.2, random_state=20)


In [171]:
# standalone support vector machine
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

In [173]:
score = cross_val_score(SVC(),X,y,cv = 5)
score.mean()

0.6906445672191528

In [179]:
# Use bagging now with svm
from sklearn.ensemble import BaggingClassifier
bag_model = BaggingClassifier(estimator=SVC(),
                              n_estimators=100,
                              max_samples=0.80,
                              random_state = 0
                             )
score = cross_val_score(bag_model,X,y,cv = 5)
score.mean()

0.6839656230816453

As you can see above, using bagging in case of SVM doesn't make much difference in terms of model accuracy. Bagging is effective when we have high variance and instable model such as decision tree. Let's explore how bagging changes the performance for a decision tree classifier.

## 6. Train a model using decision tree and then using bagging

In [185]:
from sklearn.tree import DecisionTreeClassifier
score = cross_val_score(DecisionTreeClassifier(),X,y,cv=5)
score.mean()

0.7205279312461633

In [189]:
# Use bagging now with decision tree
bag_model = BaggingClassifier(estimator=DecisionTreeClassifier(),
                             n_estimators=100,
                             max_samples=.80,
                             random_state=0
                             )
score = cross_val_score(bag_model,X,y,cv=5)
score.mean()

0.7959422958870472

You can see that with bagging the score improved from 72.05% to 79.59%

## Train a model using Random Forest which itself uses bagging underneath

In [194]:
from sklearn.ensemble import RandomForestClassifier

In [196]:
score = cross_val_score(RandomForestClassifier(),X,y,cv = 5)
score.mean()

0.8159545733578885

#### Random forest gave even a better performance with 81.7% as score. Underneath it used bagging where it sampled not only data rows but also the columns (or features)